pkg/tss/tpm_linux.go

```diff
@@ -0,0 +1,260 @@
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package tss
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/google/go-tpm/tpm"
+	"github.com/google/go-tpm/tpm2"
+)
+
+const (
+	tpmRoot = "/sys/class/tpm"
+
+	tpmPtManufacturer = 0x00000100 + 5  // PT_FIXED + offset of 5
+	tpmPtVendorString = 0x00000100 + 6  // PT_FIXED + offset of 6
+	tpmPtFwVersion1   = 0x00000100 + 11 // PT_FIXED + offset of 11
+)
+
+// TCGVendorID TPM manufacturer id
+type TCGVendorID uint32
+
+// TPMVersion is used to configure a preference in
+// which TPM to use, if multiple are available.
+type TPMVersion uint8
+
+// TPMInterface indicates how the client communicates
+// with the TPM.
+type TPMInterface uint8
+
+var vendors = map[TCGVendorID]string{
+	1095582720: "AMD",
+	1096043852: "Atmel",
+	1112687437: "Broadcom",
+	1229081856: "IBM",
+	1213220096: "HPE",
+	1297303124: "Microsoft",
+	1229346816: "Infineon",
+	1229870147: "Intel",
+	1279610368: "Lenovo",
+	1314082080: "National Semiconductor",
+	1314150912: "Nationz",
+	1314145024: "Nuvoton Technology",
+	1363365709: "Qualcomm",
+	1397576515: "SMSC",
+	1398033696: "ST Microelectronics",
+	1397576526: "Samsung",
+	1397641984: "Sinosun",
+	1415073280: "Texas Instruments",
+	1464156928: "Winbond",
+	1380926275: "Fuzhou Rockchip",
+	1196379975: "Google",
+}
+
+// TPM versions
+const (
+	TPMVersionAgnostic TPMVersion = iota
+	TPMVersion12
+	TPMVersion20
+)
+
+// TPM interfaces
+const (
+	TPMInterfaceDirect TPMInterface = iota
+	TPMInterfaceKernelManaged
+	TPMInterfaceDaemonManaged
+)
+
+// OpenConfig encapsulates settings passed to OpenTPM().
+type OpenConfig struct {
+	// TPMVersion indicates which TPM version the library should
+	// attempt to use. If the specified version is not available,
+	// ErrTPMNotAvailable is returned. Defaults to TPMVersionAgnostic.
+	TPMVersion TPMVersion
+}
+
+// TPM interfaces with a TPM device on the system.
+type TPM struct {
+	version TPMVersion
+	interf  TPMInterface
+
+	sysPath string
+	rwc     io.ReadWriteCloser
+}
+
+// probedTPM identifies a TPM device on the system, which
+// is a candidate for being used.
+type probedTPM struct {
+	Version TPMVersion
+	Path    string
+}
+
+// TPMInfo contains information about the version & interface
+// of an open TPM.
+type TPMInfo struct {
+	Version      TPMVersion
+	Interface    TPMInterface
+	VendorInfo   string
+	Manufacturer TCGVendorID
+
+	// FirmwareVersionMajor and FirmwareVersionMinor describe
+	// the firmware version of the TPM, but are only available
+	// for TPM 2.0 devices.
+	FirmwareVersionMajor int
+	FirmwareVersionMinor int
+}
+
+func (id TCGVendorID) String() string {
+	return vendors[id]
+}
+
+func readTPM12VendorAttributes(rwc io.ReadWriter) (TPMInfo, error) {
+	var vendorInfo string
+
+	_, err := tpm.GetManufacturer(rwc)
```

Comment:
return value unused?

Reply:
vendorInfo never populated

Reply:
Yes. Not possible yet. We will add it later